File: Kotlin/Basics/DataTypesMethods.kt
package Basics

import java.lang.Math.pow
import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.roundToInt
import kotlin.math.sqrt

fun main(){
    // ============================
    // String methods
    // ============================
    val myStr = "Hello, Kotlin! "
    println("""
        Length of a string: ${myStr.length}
        Lowercase: ${myStr.lowercase()}
        Uppercase: ${myStr.uppercase()}
        Capitalize deprecated: ${myStr.capitalize()}
        Modern capitalize: ${myStr.replaceFirstChar { it.uppercase() }}
        Trim the last space: ${myStr.trim()}
        Trim a specific side: ${myStr.trimStart()} or ${myStr.trimEnd()}
        Substring: ${myStr.substring(0, 5)}
        Return the first n numbers: ${myStr.take(5)}
        Remove the n numbers from right to left: ${myStr.drop(7)}
        Verify if a string contains another: $myStr contains 'Hello'? ${myStr.contains("Hello")}
        Verify if a string starts with another: $myStr starts with 'He'? ${myStr.startsWith("He")}
        Verify if a string ends with another: ${myStr.trim()} ends with 'lin!'? ${myStr.trim().endsWith("lin!")}
        Replace parts of a string: Original: $myStr to ${myStr.replace("Kotlin", "Android")}
        More complex: Original: $myStr to ${myStr.replace(myStr, "I'm Dialcadev:)")}
        Split parts of a string and transform in a list or array: ${myStr.split(" ")}
    """.trimIndent())
    // =======================
    // Number methods
    // ======================
    val x: Int = 17
    val pi: Double = 3.14159
    println("""
        Int to Double: ${x.toDouble()}
        Int to string: ${x.toString()}
        Int to char: ${x.toChar()}
        Double to int: ${pi.toInt()}
        Round double to int: ${pi.roundToInt()}
        Round to the greater number: ${ceil(pi).toInt()}
        Round to the nearest to down number: ${floor(pi).toInt()}
    """.trimIndent())
    // ====================
    // Math operations
    // ===================
    println("""
        Absolute value: ${abs(-5)}
        Max value between two numbers: ${max(3, 7)}
        Min value between two numbers: ${min(17, 99)}
        Square root: ${sqrt(16.00)}
        Potentiation: ${2.0.pow(3.0)}
    """.trimIndent())
    // ==================
    // Char methods
    // =================
    val myChar: Char = 'D'
    println("""
        Verify if the char is a digit: $myChar is a digit? ${myChar.isDigit()}
        Verify if the char is a letter: $myChar is a letter? ${myChar.isLetter()}
        Verify if the chat is a whiteSpace: $myChar is a white space?? ${myChar.isWhitespace()}
        Lower and Upper case: Lower ${myChar.lowercaseChar()}, Upper ${myChar.uppercaseChar()}
    """.trimIndent())
}

File: dart/basics/data_types_methods.dart
void main(){
  // ========== STRING METHODS ==========
  String myString = 'Hello, Dart!';
  print("""
  Original String: $myString
  Length: ${myString.length}
  Uppercase: ${myString.toUpperCase()}
  Lowercase: ${myString.toLowerCase()}
  Contains 'Dart' (case-sensitive) : ${myString.contains('Dart')}
  Is Empty: ${myString.isEmpty}
  Is not Empty: ${myString.isNotEmpty}
  Trims: '${myString.trim()}'
  Trim left: '${myString.trimLeft()}'
  Trim right: '${myString.trimRight()}'
  Replace 'Dart' with 'Flutter': ${myString.replaceAll('Dart', 'Flutter')}
  Replace first 'l' with 'L': ${myString.replaceFirst('l', 'L')}
  Split by space: ${myString.split(' ')}\
  Substring (0,5) (Should know the exact length to avoid errors): ${myString.substring(0,5)}
  'Dart' starts with 'Da': ${myString.startsWith('Da')}
  'Dart' ends with 'rt': ${myString.endsWith('rt')}
  Index of 'Dart': ${myString.indexOf('Dart')}
  Last index of 'l': ${myString.lastIndexOf('l')}
  Compare to 'Hello, dart!' (0 means equal, negative means less, positive means greater): ${myString.compareTo('Hello, dart!')}
  Pad left to length 20 with '*': '${myString.padLeft(20, '*')}'
  Pad right to length 20 with '-': '${myString.padRight(20, '-')}'
  """);
  // ========== NUMBER METHODS ==========
  var myNum = 17;
  var myDouble = 3.14;
  print("""
  Original Number: $myNum
  Is Even (Just for integers): ${myNum.isEven}
  Is Odd (Just for integers): ${myNum.isOdd}
  Is Infinite: ${myNum.isInfinite}
  Is NaN: ${myNum.isNaN}
  Is Finite: ${myNum.isFinite}
  Is Negative: ${myNum.isNegative}
  Double Value: $myDouble
  Double to Int (Truncated, removes decimal part): ${myDouble.toInt()}
  To Double: ${myNum.toDouble()}
  To String (Just for integers): ${myNum.toString()}
  Round (Most close integer, more accurate): ${myDouble.round()}
  Floor (Lower integer): ${myDouble.floor()}
  Ceil (Higher integer): ${myDouble.ceil()}
  Truncate (Removes decimal part, useful for removing decimals without rounding): ${myDouble.truncate()}
  Abs Value: ${(-myNum).abs()}
  Signum (1 if positive, -1 if negative, 0 if zero): ${myNum.sign}
  Compare to 10 (0 means equal, negative means less, positive means greater): ${myNum.compareTo(10)}
  Clamp between 10 and 20 (Keeps the number within the range): ${myNum.clamp(10, 20)}
  Parse from String '42': ${int.parse('42')}
  Parse from String '3.14': ${double.parse('3.14')}
  Try Parse from String 'not a number' (Returns null if parsing fails): ${int.tryParse('not a number')}
  """);
}